### PR TITLE
Extract metrics API that is decoupled from dropwizard [PLEN-28]

### PR DESCRIPTION
**Pull request by nicu-da**
This will allow us to add another implementation for metrics in the future (OpenTelemetry) while keeping the API unchanged.
In a follow up PR some changes to the API will be done to add support for labels in the API itself.

### CODE REVIEW

**Comment by atriantafyllos-da** [comment]
As a first comment this will break canton. I must try to compile with canton to see the conflicts.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala [hunk 1]
```diff
@@ -24,111 +21,166 @@ object MetricHandle {
...
-
-    def timer(name: MetricName): Timer = Timer(name, registry.timer(name))
+    def timer(name: MetricName): Timer
 
-    def varGauge[T](name: MetricName, initial: T): VarGauge[T] =
-      addGauge(name, Gauges.VarGauge[T](initial), _.updateValue(initial))
+    def gauge[T](name: MetricName, initial: T): Gauge[T]
 
-    def addGauge[T <: codahale.Gauge[M], M](
+    def gaugeWithSupplier[T](
```
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:28]: removing `addGauge` will break canton in 
https://github.com/DACH-NY/canton/blob/e888db5dc5ab7dccba61ebef171fb2e5c46b4d3d/community/common/src/main/scala/com/digitalasset/canton/metrics/MetricHandle.scala#L30-L36

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala [hunk 2]
```diff
@@ -24,111 +21,166 @@ object MetricHandle {
...
-    def update(duration: Long, unit: TimeUnit): Unit = metric.update(duration, unit)
-    def getCount: Long = metric.getCount
-    def getSnapshot: Snapshot = metric.getSnapshot
-    def getMeanRate: Double = metric.getMeanRate
-    def time(): Context = metric.time()
+  object Timer {
+    type TimerStop = () => Unit
   }
 
-  sealed case class Gauge[U <: codahale.Gauge[T], T](name: String, metric: U)
```
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:98]: This change also breaks canton e.g. https://github.com/DACH-NY/canton/blob/e888db5dc5ab7dccba61ebef171fb2e5c46b4d3d/community/common/src/main/scala/com/digitalasset/canton/crypto/store/db/DbCryptoPrivateStore.scala#L70.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala [hunk 3]
```diff
@@ -24,111 +21,166 @@ object MetricHandle {
...
-      extends MetricHandle[codahale.Histogram] {
+  sealed trait Histogram extends MetricHandle {
+
     def metricType: String = "Histogram"
+    def update(value: Long): Unit
+    def update(value: Int): Unit
 
-    def update(value: Long): Unit = metric.update(value)
-    def update(value: Int): Unit = metric.update(value)
-    def getSnapshot: Snapshot = metric.getSnapshot
```
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:129]: Needed by canton.

File: ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala [hunk 4]
```diff
@@ -57,31 +53,24 @@ object Timed {
   }
 
   def future[T](timer: Timer, future: => Future[T]): Future[T] = {
-    this.future[T](timer.metric, future)
-  }
-
-  def future[T](timer: codahale.Timer, future: => Future[T]): Future[T] = {
```
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala:63]: this function must be exported for canton

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala [hunk 5]
```diff
@@ -24,111 +21,166 @@ object MetricHandle {
...
       new DatabaseMetrics(prefix, name, registry)
   }
 
-  sealed case class Timer(name: String, metric: codahale.Timer)
-      extends MetricHandle[codahale.Timer] {
+  sealed trait Timer extends MetricHandle {
+
     def metricType: String = "Timer"
 
-    def timeEitherT[E, A](ev: EitherT[Future, E, A]): EitherT[Future, E, A] = {
```
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:87]: breaks canton if removed

**Review by atriantafyllos-da — CHANGES_REQUESTED** [review]

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala (hunk 1 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:28]: Based on what I'm seeing in canton, the `RefGauge` is more or less a `VarGauge`. The metrics using `RefGauge` should in fact use a `gaugeWithSupplier`, and the `RefGuage` should be deleted.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala (hunk 2 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:98]: We're mixing concepts in a pretty difficult way here. The `TimedLoadGoage` should not exist, instead this should be an extension for `Timer` that adds cats ways to work with cats.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala (hunk 3 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:129]: We should not depend on the internals of the metrics themselves. If required for testing we can create assertions that work with metrics.

File: ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala (hunk 4 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala:63]: It depends on `Dropwizard`, canton should not use dropwizard metrics directly.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala (hunk 5 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:87]: It should be moved together with themethods from the `TimedLoadGauge` as it seems to target the same API.

**Comment by nicu-da** [comment]
> As a first comment this will break canton. I must try to compile with canton to see the conflicts.

@atriantafyllos-da Thank you very much for checking the integration in canton. I think you discovered some good spots where it would break, and based on what I've seen from the usages we removed I think the changes should happen in canton itself, so adapt to the simplified API.

**Review by simonmaxen-da — COMMENTED** [review]
LGTM

**Review by atriantafyllos-da — APPROVED** [review]
Approved but please do not merge before the PR for canton is approved.

**Pull request merged**